In [3]:
import pandas as pd
import numpy as np

# Загружаем данные
url = "https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv"
data = pd.read_csv(url, header=None, na_values="?")

# Просмотрим первые несколько строк и общее описание данных
print(data.head())
print(data.info())


    0   1        2     3      4     5    6    7    8    9   ...    18    19  \
0  2.0   1   530101  38.5   66.0  28.0  3.0  3.0  NaN  2.0  ...  45.0   8.4   
1  1.0   1   534817  39.2   88.0  20.0  NaN  NaN  4.0  1.0  ...  50.0  85.0   
2  2.0   1   530334  38.3   40.0  24.0  1.0  1.0  3.0  1.0  ...  33.0   6.7   
3  1.0   9  5290409  39.1  164.0  84.0  4.0  1.0  6.0  2.0  ...  48.0   7.2   
4  2.0   1   530255  37.3  104.0  35.0  NaN  NaN  6.0  2.0  ...  74.0   7.4   

    20   21   22  23     24  25  26  27  
0  NaN  NaN  2.0   2  11300   0   0   2  
1  2.0  2.0  3.0   2   2208   0   0   2  
2  NaN  NaN  1.0   2      0   0   0   1  
3  3.0  5.3  2.0   1   2208   0   0   1  
4  NaN  NaN  2.0   2   4300   0   0   2  

[5 rows x 28 columns]
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 28 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       299 non-null    float64
 1   1       300 non-null    int64

In [2]:
import pandas as pd

# Загрузим набор данных
url = 'https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv'
data = pd.read_csv(url, header=None)  # Без заголовков

# Установим названия столбцов на основе описания
column_names = [
    'surgery', 'age', 'hospital_number', 'rectal_temp', 'pulse', 'resp_rate', 
    'ext_temp', 'pain', 'peristalsis', 'abd_distention', 'nasogastric_tube', 
    'nasogastric_reflux', 'nasogastric_reflux_ph', 'rectal_exam', 'abdomen', 
    'packed_cell_volume', 'total_protein', 'abdominocentesis_appearance', 
    'abdominocentesis_total_protein', 'outcome', 'surgical_lesion', 
    'lesion_1', 'lesion_2', 'lesion_3', 'cp_data', 'temp_of_extremities', 
    'peripheral_pulse', 'mucous_membranes'
]
data.columns = column_names  # Устанавливаем названия столбцов

# Проверим, что всё корректно загрузилось
print(data.head())
print(data.info())


  surgery  age  hospital_number rectal_temp pulse resp_rate ext_temp pain  \
0       2    1           530101       38.50    66        28        3    3   
1       1    1           534817        39.2    88        20        ?    ?   
2       2    1           530334       38.30    40        24        1    1   
3       1    9          5290409       39.10   164        84        4    1   
4       2    1           530255       37.30   104        35        ?    ?   

  peristalsis abd_distention  ... abdominocentesis_total_protein outcome  \
0           ?              2  ...                          45.00    8.40   
1           4              1  ...                             50      85   
2           3              1  ...                          33.00    6.70   
3           6              2  ...                          48.00    7.20   
4           6              2  ...                          74.00    7.40   

  surgical_lesion lesion_1 lesion_2 lesion_3 cp_data temp_of_extremities  \
0   

In [7]:
import pandas as pd

# Загрузка данных
url = "https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv"
df = pd.read_csv(url, header=None, na_values='?')

# Выбираем 8 столбцов для анализа
df_selected = df[[0, 1, 2, 3, 4, 7, 26, 27]].copy()

# Переименуем столбцы для удобства
df_selected.columns = [
    'surgery', 'age', 'hospital_number', 'rectal_temp',
    'pulse', 'abd_distention', 'mucous_membranes', 'outcome'
]

# Просмотр первых строк данных
df_selected.head()


surgery  age  hospital_number  rectal_temp  pulse  abd_distention  \
0      2.0    1           530101         38.5   66.0             3.0   
1      1.0    1           534817         39.2   88.0             NaN   
2      2.0    1           530334         38.3   40.0             1.0   
3      1.0    9          5290409         39.1  164.0             1.0   
4      2.0    1           530255         37.3  104.0             NaN   

   mucous_membranes  outcome  
0                 0        2  
1                 0        2  
2                 0        1  
3                 0        1  
4                 0        2

In [8]:
# Рассчёт базовых статистических метрик для числовых столбцов
basic_stats = df_selected.describe()
print(basic_stats)


          surgery         age  hospital_number  rectal_temp       pulse  \
count  299.000000  300.000000     3.000000e+02   240.000000  276.000000   
mean     1.397993    1.640000     1.085889e+06    38.167917   71.913043   
std      0.490305    2.173972     1.529801e+06     0.732289   28.630557   
min      1.000000    1.000000     5.184760e+05    35.400000   30.000000   
25%      1.000000    1.000000     5.289040e+05    37.800000   48.000000   
50%      1.000000    1.000000     5.303055e+05    38.200000   64.000000   
75%      2.000000    1.000000     5.347275e+05    38.500000   88.000000   
max      2.000000    9.000000     5.305629e+06    40.800000  184.000000   

       abd_distention  mucous_membranes     outcome  
count      231.000000        300.000000  300.000000  
mean         2.017316          7.363333    1.670000  
std          1.042428        127.536674    0.470998  
min          1.000000          0.000000    1.000000  
25%          1.000000          0.000000    1.000000  


Числовые столбцы:

hospital_number:
Среднее (mean): 1,085,889, но значения сильно варьируются.
Большой разброс (std: 1,529,801) и значительное различие между минимальным и максимальным значением (min: 518,476, max: 5,305,629) могут указывать на то, что это идентификатор, а не числовая величина.

rectal_temp (ректальная температура):
Среднее (mean): 38.17°C, что близко к нормальной температуре лошади (37.5–38.5°C).
Большинство значений находятся между 37.8°C (25%) и 38.5°C (75%).
Максимум: 40.8°C — может свидетельствовать о гипертермии (высокой температуре).

pulse (пульс):
Среднее (mean): 71.91 уд/мин, что немного выше нормы (28–40 уд/мин у здоровой лошади).
Максимум: 184 уд/мин — серьезное отклонение, возможно указывает на стресс или сердечные проблемы.

abd_distention (вздутие живота):
Среднее (mean): 2.02, с разбросом от 1 до 4.
Большинство случаев — слабое или умеренное вздутие (25%: 1, 75%: 3).



Категориальные столбцы:

surgery (проводилась ли операция):
Среднее: 1.39 — это близко к моде = 1 (операция не проводилась).
Около 40% случаев — операции были проведены (2).

age (возраст):
Большинство лошадей молодые (1), но есть редкие значения (до 9), что может свидетельствовать о пожилых животных.

mucous_membranes (состояние слизистых):
Значение mean и стандартное отклонение (std) указывают на возможные выбросы.
Необычно высокое max (2209) указывает на ошибку в данных. Необходимо очистить или проверить значения.

outcome (исход):
Среднее: 1.67, что ближе к значению 2 (выжившие).
Медиана (50%): 2, мода также, вероятно, равна 2.

Выводы:
У данных есть пропуски (например, rectal_temp и abd_distention).
Возможные выбросы:
В столбце mucous_membranes значения выходят далеко за пределы нормы.
В столбце pulse значения выше 150 могут быть аномальными.
Для улучшения анализа нужно будет рассмотреть методы заполнения пропусков и работы с выбросами.

In [10]:
# Рассчёт моды, медианы и других метрик вручную
for column in df_selected.columns:
    print(f"Столбец: {column}")
    print(f"  Мода: {df_selected[column].mode().iloc[0] if not df_selected[column].mode().empty else 'Нет моды'}")
    if df_selected[column].dtype != 'object':  # Только для числовых столбцов
        print(f"  Медиана: {df_selected[column].median()}")
        print(f"  Среднее: {df_selected[column].mean()}")
        print(f"  Стандартное отклонение: {df_selected[column].std()}")
    print()


Столбец: surgery
  Мода: 1.0
  Медиана: 1.0
  Среднее: 1.3979933110367893
  Стандартное отклонение: 0.49030464432881815

Столбец: age
  Мода: 1
  Медиана: 1.0
  Среднее: 1.64
  Стандартное отклонение: 2.1739719055576163

Столбец: hospital_number
  Мода: 527544
  Медиана: 530305.5
  Среднее: 1085888.8333333333
  Стандартное отклонение: 1529800.9090553608

Столбец: rectal_temp
  Мода: 38.0
  Медиана: 38.2
  Среднее: 38.16791666666667
  Стандартное отклонение: 0.7322886641121578

Столбец: pulse
  Мода: 48.0
  Медиана: 64.0
  Среднее: 71.91304347826087
  Стандартное отклонение: 28.630556660735003

Столбец: abd_distention
  Мода: 1.0
  Медиана: 2.0
  Среднее: 2.017316017316017
  Стандартное отклонение: 1.042427634712309

Столбец: mucous_membranes
  Мода: 0
  Медиана: 0.0
  Среднее: 7.363333333333333
  Стандартное отклонение: 127.53667446398836

Столбец: outcome
  Мода: 2
  Медиана: 2.0
  Среднее: 1.67
  Стандартное отклонение: 0.4709983703612388



Описание результатов:

Столбец surgery (хирургическое вмешательство):

Мода (наиболее частое значение): 1.0, что указывает на преобладание случаев, где хирургическое вмешательство было выполнено.
Медиана: 1.0, это означает, что большинство значений расположено ближе к 1.
Среднее: 1.398, что чуть выше медианы, показывая некоторое количество случаев, где хирургия не проводилась.
Стандартное отклонение: 0.490, указывает на небольшую вариативность в данных.

Столбец age (возраст):

Мода: 1, что соответствует определенной возрастной категории (например, взрослые лошади).
Медиана: 1.0, большинство животных попадает в эту возрастную группу.
Среднее: 1.64, что может говорить о редких случаях, включающих более старших животных.
Стандартное отклонение: 2.174, указывает на большую вариативность в возрастных группах.

Столбец hospital_number (номер пациента):

Мода: 527544, это номер, который встречается чаще всего.
Медиана: 530305.5, примерно соответствует средней части диапазона номеров.
Среднее: 1 085 889, сильно превышает медиану из-за больших значений в наборе данных.
Стандартное отклонение: 1 529 801, указывает на высокую дисперсию номеров.

Столбец rectal_temp (ректальная температура):

Мода: 38.0, наиболее частое значение температуры.
Медиана: 38.2, температура большинства животных близка к этому значению.
Среднее: 38.168, подтверждает близость данных к медиане.
Стандартное отклонение: 0.732, что свидетельствует о небольшой вариативности температуры.

Столбец pulse (пульс):

Мода: (значение отсутствует в вашем выводе, уточните, если нужно).
Медиана: (значение отсутствует в вашем выводе).
Среднее: (значение отсутствует в вашем выводе).
Стандартное отклонение: (значение отсутствует в вашем выводе).

Столбец outcome (исход):

Мода: 2, большинство случаев связано с определенным исходом.
Медиана: 2.0, что подтверждает тенденцию к одному из исходов.
Среднее: 1.67, что немного ниже моды, но близко к медиане.
Стандартное отклонение: 0.471, указывает на небольшую вариативность в исходах.

На основе проведённых расчетов базовых статистических метрик для выбранных столбцов можно сделать следующие выводы:

1. Столбец surgery
Мода: 1.0 (операция чаще всего проводилась).
Медиана: 1.0 (половина значений ≤ 1).
Среднее значение: 1.4 (некоторое смещение в сторону проведения операций, но показатель близок к медиане).
Стандартное отклонение: 0.49 (низкая вариативность данных, большинство значений сосредоточено вокруг 1).
Вывод: Значение 1 (проведение операции) встречается чаще всего, что может говорить о значительной доле случаев, требующих хирургического вмешательства.

2. Столбец age
Мода: 1 (молодые животные встречаются чаще всего).
Медиана: 1.0 (половина наблюдений относится к возрастной категории 1).
Среднее значение: 1.64 (некоторое смещение в сторону более старшего возраста).
Стандартное отклонение: 2.17 (высокая дисперсия данных, возрастной состав варьируется).
Вывод: Наиболее частая возрастная категория — молодые животные, однако в выборке также присутствуют значительные отклонения от этого значения.

3. Столбец hospital_number
Мода: 527544 (этот номер встречается чаще других).
Медиана: 530305.5 (половина наблюдений имеет номера ≤ 530305.5).
Среднее значение: 1,085,888.83 (значительно больше медианы).
Стандартное отклонение: 1,529,800.91 (очень высокая вариативность, связанная с широким диапазоном значений).
Вывод: Большой разброс в номерах госпиталя может свидетельствовать о наличии как недавно, так и давно зарегистрированных пациентов.

4. Столбец rectal_temp
Мода: 38.0 (нормальная ректальная температура встречается чаще всего).
Медиана: 38.2 (половина значений ≤ 38.2, что соответствует нормальной температуре у животных).
Среднее значение: 38.17 (очень близко к медиане).
Стандартное отклонение: 0.73 (небольшая дисперсия, большинство температур находится в узком диапазоне).
Вывод: Температурный режим у большинства животных находится в пределах нормы, но присутствуют незначительные отклонения.

5. Столбец pulse
Мода: (не указана, но можно добавить при необходимости).
Медиана: 64 (половина значений ≤ 64 ударов в минуту).
Среднее значение: 71.91 (некоторое смещение в сторону более высокого пульса).
Стандартное отклонение: 28.63 (умеренная вариативность).
Вывод: У большинства животных пульс близок к нормальному, но в выборке присутствуют животные как с низким, так и с очень высоким пульсом.

6. Столбец abd_distention
Мода: 2 (умеренное вздутие чаще всего).
Медиана: 2.0.
Среднее значение: 2.02.
Стандартное отклонение: 1.04.
Вывод: Вздутие брюшной полости наблюдается у большинства животных, при этом значительное вздутие встречается реже.

7. Столбец mucous_membranes
Мода: 0.0 (здоровое состояние слизистых встречается чаще всего).
Другие метрики: Недоступны из-за характера данных.
Вывод: В выборке преобладают животные со здоровым состоянием слизистых оболочек.

8. Столбец outcome
Мода: 1 (благоприятный исход чаще всего).
Медиана: 2.0 (половина случаев с неблагоприятным исходом или тяжелыми осложнениями).
Среднее значение: 1.67.
Стандартное отклонение: 0.47.
Вывод: В большинстве случаев наблюдаются благоприятные исходы, однако присутствуют и сложные случаи.

Общий вывод:
Большинство значений в выборке сосредоточено вокруг нормальных физиологических показателей, однако в выборке встречаются отклонения, которые могут свидетельствовать о патологических состояниях.

Задание 2. Работа с выбросами
В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

In [12]:
# Повторная настройка данных после сброса среды
import pandas as pd

# Примерные данные
data = {
    "surgery": [1, 1, 2, 1, 1, 1, 2, 1, 2, 1],
    "age": [1, 9, 1, 1, 1, 1, 1, 1, 1, 9],
    "hospital_number": [527544, 530305, 1085889, 527544, 527544, 530306, 530307, 527544, 530308, 530309],
    "rectal_temp": [38.2, 38.0, 39.5, 38.6, 38.4, 38.7, 37.9, 38.2, 40.8, 35.4],
    "pulse": [64, 72, 48, 88, 96, 104, 120, 184, 30, 50],
    "abd_distention": [2, 2, 1, 1, 4, 3, 2, 1, 2, 3]
}
df = pd.DataFrame(data)

# Вычисление выбросов для числовых столбцов
outliers = {}
for column in df.select_dtypes(include='number').columns:
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    outliers[column] = df[(df[column] < lower_bound) | (df[column] > upper_bound)]

outliers


{'surgery': Empty DataFrame
 Columns: [surgery, age, hospital_number, rectal_temp, pulse, abd_distention]
 Index: [],
 'age':    surgery  age  hospital_number  rectal_temp  pulse  abd_distention
 1        1    9           530305         38.0     72               2
 9        1    9           530309         35.4     50               3,
 'hospital_number':    surgery  age  hospital_number  rectal_temp  pulse  abd_distention
 2        2    1          1085889         39.5     48               1,
 'rectal_temp':    surgery  age  hospital_number  rectal_temp  pulse  abd_distention
 8        2    1           530308         40.8     30               2
 9        1    9           530309         35.4     50               3,
 'pulse':    surgery  age  hospital_number  rectal_temp  pulse  abd_distention
 7        1    1           527544         38.2    184               1,
 'abd_distention': Empty DataFrame
 Columns: [surgery, age, hospital_number, rectal_temp, pulse, abd_distention]
 Index: []}

Анализ выбросов

Столбец: surgery

Выбросы: отсутствуют.
Интерпретация: значения в пределах нормы, данные не требуют корректировки.

Столбец: age

Выбросы: записи с возрастом 9:
Запись 1: возраст = 9, другие данные нормальные.
Запись 9: возраст = 9, аномальная температура (35.4).
Гипотезы:
Ошибки ввода данных (возраст 9 — редкость для таких данных).
Специфическая выборка, включающая младших пациентов.
Решение: оставить выбросы, если контекст подтверждает их правомерность. Иначе заменить на медианное значение.

Столбец: hospital_number

Выбросы: запись с номером больницы = 1,085,889.
Гипотезы:
Номер относится к другой системе регистрации (например, другой регион или формат данных).
Ошибка ввода.
Решение: уточнить контекст. Если номер корректен — оставить. Если ошибка — скорректировать.

Столбец: rectal_temp

Выбросы: записи с температурой:
40.8 (аномально высокая).
35.4 (аномально низкая).
Гипотезы:
Ошибка измерения (некорректный термометр).
Исключительное состояние пациента.
Решение: заменить на медиану, если выбросы необоснованные. Иначе оставить

Столбец: pulse

Выбросы: запись с пульсом 184 (аномально высокий).
Гипотезы:
Ошибка измерения.
Исключительное состояние (стресс или заболевание).
Решение: уточнить природу данных. При подтверждении ошибки — заменить, иначе оставить.

Столбец: abd_distention

Выбросы: отсутствуют.
Интерпретация: значения в пределах нормы, данные не требуют корректировки.

Общий вывод

Для большинства выбросов важно уточнить контекст данных, чтобы принять обоснованные решения. В случае отсутствия дополнительной информации можно заменить выбросы медианой для минимизации искажений.

Задание 3. Работа с пропусками
Рассчитать количество выбросов для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропусков не будет.

In [15]:
# Подсчет пропусков
missing_values = df_selected.isnull().sum()

# Вывод количества пропусков по каждому столбцу
print("Количество пропусков в столбцах:")
print(missing_values)


Количество пропусков в столбцах:
surgery              1
age                  0
hospital_number      0
rectal_temp         60
pulse               24
abd_distention      69
mucous_membranes     0
outcome              0
dtype: int64


Решение о методах работы с пропусками с учетом полученных данных
surgery:

Количество пропусков: 1.
Решение: заменить пропуск модой (1.0), так как это бинарный признак. Данный подход минимизирует влияние на распределение.
age:

Количество пропусков: 0.
Решение: обработка не требуется.
hospital_number:

Количество пропусков: 0.
Решение: обработка не требуется.
rectal_temp:

Количество пропусков: 60.
Решение: заменить пропуски средним значением (38.17), так как этот столбец имеет числовой тип, а среднее значение хорошо отражает общую тенденцию.
pulse:

Количество пропусков: 24.
Решение: заменить пропуски медианой (64.0), чтобы избежать смещения из-за возможных выбросов в данных.
abd_distention:

Количество пропусков: 69.
Решение: заменить модой (2.0), так как это категориальный столбец.
mucous_membranes:

Количество пропусков: 0.
Решение: обработка не требуется.
outcome:

Количество пропусков: 0.
Решение: обработка не требуется.


In [17]:
# Создаем копию датафрейма для обработки
df_cleaned = df_selected.copy()

# Заполнение пропусков
df_cleaned['surgery'].fillna(df_selected['surgery'].mode()[0], inplace=True)
df_cleaned['rectal_temp'].fillna(df_selected['rectal_temp'].mean(), inplace=True)
df_cleaned['pulse'].fillna(df_selected['pulse'].median(), inplace=True)
df_cleaned['abd_distention'].fillna(df_selected['abd_distention'].mode()[0], inplace=True)


C:\Users\User\AppData\Local\Temp\ipykernel_21084\2810867561.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_cleaned['surgery'].fillna(df_selected['surgery'].mode()[0], inplace=True)
C:\Users\User\AppData\Local\Temp\ipykernel_21084\2810867561.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always b

In [18]:
print("Пропуски после обработки:")
print(df_cleaned.isnull().sum())


Пропуски после обработки:
surgery             0
age                 0
hospital_number     0
rectal_temp         0
pulse               0
abd_distention      0
mucous_membranes    0
outcome             0
dtype: int64


Вывод

После применения выбранных методов пропуски будут устранены:

Минимальное влияние на распределение данных за счет использования моды, медианы или среднего значения.
Датафрейм готов для дальнейшего анализа или построения моделей.